src/github.com/matrix-org/dendrite/clientapi/auth/tokens/tokens_handlers.go

```diff
@@ -0,0 +1,115 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tokens
+
+import (
+	"errors"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/matrix-org/dendrite/clientapi/jsonerror"
+	"github.com/matrix-org/util"
+)
+
+// GetUserFromToken returns the user associated with the token
+// Returns the error if something goes wrong.
+// Warning: Does not validate the token. Use ValidateToken for that.
+func GetUserFromToken(token string) (user string, err error) {
+	mac, err := deSerializeMacaroon(token)
+	if err != nil {
+		return
+	}
+
+	user = string(mac.Id()[:])
+	return
+}
+
+// ValidateToken validates that the Token is understood and was signed by this server.
+// Returns nil if token is valid, otherwise returns a error response which can be sent to client.
+func ValidateToken(op TokenOptions, token string) *util.JSONResponse {
+	macaroon, err := deSerializeMacaroon(token)
+	if err != nil {
+		return &util.JSONResponse{
+			Code: http.StatusUnauthorized,
+			JSON: jsonerror.UnknownToken("Token does not represent a valid macaroon"),
+		}
+	}
+
+	caveats, err := macaroon.VerifySignature(op.ServerPrivateKey, nil)
+	if err != nil {
+		return &util.JSONResponse{
+			Code: http.StatusUnauthorized,
+			JSON: jsonerror.UnknownToken("Provided token was not issued by this server"),
+		}
+	}
+
+	err = verifyCaveats(caveats, op.UserID)
+	if err != nil {
+		return &util.JSONResponse{
+			Code: http.StatusForbidden,
+			JSON: jsonerror.Forbidden("Provided token not authorized"),
+		}
+	}
+	return nil
+}
+
+// verifyCaveats verifies caveats associated with a login token macaroon.
+// which are "gen = 1", "user_id = ...", "time < ..."
+// Returns nil on successful verification, else returns an error.
+func verifyCaveats(caveats []string, userID string) error {
+	// variable verified represents a bitmap
```

Comment:
This is certainly one way to do it :) Since we only really need to differentiate between good versus bad caveats, I'd say this approach works fine.